batcheval: add range tombstone support for `DeleteRange`

erikgrinaker opened the pull request:
This patch adds the parameter `UseExperimentalRangeTombstone` for
`DeleteRange`, which deletes the span using an MVCC range tombstone.
The new version gate `MVCCRangeTombstones` must be checked before using
it. `storage.ExperimentalMVCCDeleteRangeUsingTombstone()` is added to
carry out the actual deletion.

This is a bare-bones implementation to allow writing range keys via the
KV API for testing and development purposes. It has significant
shortcomings, and will be fleshed out at a later time.

Touches #70415.
Replaces #76203.

Release note: None

erikgrinaker committed:
batcheval: add MVCC range tombstone support for `DeleteRange`

This patch adds the parameter `UseExperimentalRangeTombstone` for
`DeleteRange`, which deletes the span using an MVCC range tombstone.
The new version gate `MVCCRangeTombstones` must be checked before using
it. `storage.ExperimentalMVCCDeleteRangeUsingTombstone()` is added to
carry out the actual deletion.

This is a bare-bones implementation to allow writing range keys via the
KV API for testing and development purposes. It has significant
shortcomings, and will be fleshed out at a later time.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77762)
<!-- Reviewable:end -->

aliher1911 commented on pkg/kv/kvserver/batcheval/cmd_delete_range_test.go:
```diff
@@ -0,0 +1,200 @@
+
+	// Initial data for each test: point keys at b@2, c@5, d@2, point tombstone at
+	// d@3, point intent at i@7, and a range tombstone at [k-p)@3.
+	writeInitialData := func(t *testing.T, ctx context.Context, rw storage.ReadWriter) {
```
nit, ignore if you think it doesn't help, but I find it easier to reason about test cases when I can easily map deleted range onto the key-time space
```
    //  ts
    // 7|                 i
    // 6|
    // 5|     +
    // 4|
    // 3|       -             o-----o
    // 2|   +   +
    // 1|
    // -+---------------------------------- key
    //  | a b c d e f ... i j k ... p ... z
```

erikgrinaker replied:
Yep, done.

aliher1911 commented on pkg/kv/kvserver/batcheval/cmd_delete_range_test.go:
```diff
@@ -0,0 +1,200 @@
+			returnKeys: true,
+			expectErr:  "ReturnKeys can't be used with range tombstones",
+		},
+		"intent errors with WriteIntentError": {
```
Should we test range under intent ts for completeness? Scan intent currently doesn't know about timestamp, but that's an implementation detail so might change in the future.

erikgrinaker replied:
This isn't really meant to be comprehensive, only that errors are plumbed through -- we have comprehensive tests for this in `TestMVCCHistories`, including above/below the intent timestamp (see `range_tombstone_mutations`).

aliher1911 reviewed (approved):
Is GH description correct regarding which commits belong to this PR? It looks only the last one is relevant.

Some nits as I think I already went through this code before.

erikgrinaker commented:
>Is GH description correct regarding which commits belong to this PR? It looks only the last one is relevant.

Ah, it's because I rebased the base branch `mvcc-range-tombstones` onto `master`, so the old commits that are no longer in the base branch are then considered part of this branch instead. Updated the PR description to be agnostic to the number of prior commits.

erikgrinaker commented on pkg/storage/mvcc.go:
```diff
@@ -2202,6 +2202,81 @@ func MVCCDeleteRange(
+
+	// Check for any conflicts, i.e. newer values.
+	//
+	// TODO(erikgrinaker): This introduces an O(n) read penalty. We should
```
Yeah, possibly. With stats handling (the next PR) we'll have to scan everything to update stats, but we'll eventually be implementing a fast-path when we're clearing the entire range, and in that case the TBI optimization could help with conflict checks. I'll get back to performance optimizations here later.

nicktrav reviewed (commented):
> Reviewed 20 of 20 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77762)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911 and @erikgrinaker)

___
*[pkg/storage/testdata/mvcc_histories/range_tombstone_mutations, line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77762#-MzvrNy201OM-pjYJ72M:-MzvrNy201OM-pjYJ72N:bwil8an) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d5aecb6d9616ac5107d935dac86a5692e2578f0/pkg/storage/testdata/mvcc_histories/range_tombstone_mutations#L1)):*
> ```
> # Set up some point keys, point/range tombstones, and intents.
> ```

nit: diagram would be useful here.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77762)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aliher1911, @erikgrinaker, and @nicktrav)

___
*[pkg/storage/testdata/mvcc_histories/range_tombstone_mutations, line 1 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/77762#-MzvrNy201OM-pjYJ72M:-N-j2tT638yRUhId0e31:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d5aecb6d9616ac5107d935dac86a5692e2578f0/pkg/storage/testdata/mvcc_histories/range_tombstone_mutations#L1)):*
<details><summary><i>Previously, nicktrav (Nick Travers) wrote…</i></summary><blockquote>

nit: diagram would be useful here.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
@aliher1911 May want to take a quick look at this again, since there's been a few changes to range keys since it was initially submitted -- nothing major though.

aliher1911 reviewed (approved):
I gave it another read and it looks good.

There's something wrong with the request itself though. I can see latest version with all updates in "Files changed", but neither VS-Code view, nor Reviewable shows it. Both show diff with some older commit which rolls back changes from master massively.

erikgrinaker commented:
> I gave it another read and it looks good.

Thanks!

> There's something wrong with the request itself though. I can see latest version with all updates in "Files changed", but neither VS-Code view, nor Reviewable shows it. Both show diff with some older commit which rolls back changes from master massively.

Are you sure you're using the correct base though? Remember, this used to be onto the `mvcc-range-tombstones` feature branch, but is now rebased onto master. May be that your local VSCode checkout and/or Reviewable gets that wrong and pulls in a bunch of stuff from the old base branch?

Using "Show full diff" in Reviewable shows the correct changes, as does comparing the branch to `master` in VSCode, so idk.

erikgrinaker commented:
TFTR!

bors r=aliher1911
